doc/source/gapfill.md

```diff
@@ -0,0 +1,272 @@
+:mystnb:
+:  code_prompt_show: "Show the code for opening example data"
+:  code_prompt_hide: "Hide the code for opening example data"
+
+from datetime import datetime
+
+import geoutils as gu
+import numpy as np
+import matplotlib.pyplot as plt
+
+import xdem
+
+# Load a reference DEM from 2009
+dem_2009 = xdem.DEM(xdem.examples.get_path("longyearbyen_ref_dem"), datetime=datetime(2009, 8, 1))
+# Load a DEM from 1990
+dem_1990 = xdem.DEM(xdem.examples.get_path("longyearbyen_tba_dem"), datetime=datetime(1990, 8, 1))
+# Load glacier outlines from 1990.
+glaciers_1990 = gu.Vector(xdem.examples.get_path("longyearbyen_glacier_outlines"))
+glaciers_2010 = gu.Vector(xdem.examples.get_path("longyearbyen_glacier_outlines_2010"))
+
+# Make a dictionary of glacier outlines where the key represents the associated date.
+outlines = {
+    datetime(1990, 8, 1): glaciers_1990,
+    datetime(2009, 8, 1): glaciers_2010,
+}
+
+# Cropping DEMs to a smaller extent to visualize the gap-filling better
+bounds = (dem_2009.bounds.left, dem_2009.bounds.bottom,
+          dem_2009.bounds.left + 200 * dem_2009.res[0], dem_2009.bounds.bottom + 150 * dem_2009.res[1])
+dem_2009 = dem_2009.crop(bounds)
+dem_1990 = dem_1990.crop(bounds)
+```
+
+We create a difference of DEMs object {class}`xdem.ddem.dDEM` to experiment on:
+
+```{code-cell} ipython3
+ddem = xdem.dDEM(raster=dem_2009 - dem_1990, start_time=dem_1990.datetime, end_time=dem_2009.datetime)
+
+# The example DEMs are void-free, so let's make some random voids.
+# Introduce 50000 nans randomly throughout the dDEM.
+mask = np.zeros_like(ddem.data, dtype=bool)
+mask.ravel()[(np.random.choice(ddem.data.size, int(ddem.data.size/5), replace=False))] = True
+ddem.set_mask(mask)
+```
+
+## Linear spatial interpolation
+
+Linear spatial interpolation (also often called bilinear interpolation) of dDEMs is arguably the simplest approach: voids are filled by an average of the surrounding pixels values, weighted by their distance to the void pixel.
+
+```{code-cell} ipython3
+ddem_linear = ddem.interpolate(method="linear")
+```
+
+```{code-cell} ipython3
+:tags: [hide-input]
+:mystnb:
+:  code_prompt_show: "Show plotting code"
+:  code_prompt_hide: "Hide plotting code"
+
+ddem_linear = ddem.copy(new_array=ddem_linear)
+
+# Plot before and after
+f, ax = plt.subplots(1, 2)
+ax[0].set_title("Before linear\ngap-filling")
+ddem.plot(cmap='RdYlBu', vmin=-10, vmax=10, ax=ax[0])
+ax[1].set_title("After linear\ngap-filling")
+ddem_linear.plot(cmap='RdYlBu', vmin=-10, vmax=10, ax=ax[1], cbar_title="Elevation differences (m)")
+_ = ax[1].set_yticklabels([])
+```
+
+## Local hypsometric interpolation
+
+This approach assumes that there is a relationship between the elevation and the elevation change in the dDEM, which is often the case for glaciers.
+Elevation change gradients in late 1900s and 2000s on glaciers often have the signature of large melt in the lower parts, while the upper parts might be less negative, or even positive.
```

Comment:
I would replace "melt" by thinning, as we cannot know if the thinning is due to melt or dynamic.